### ISSUE TITLE
No preview for some URLs

### ISSUE THREAD

**Issue by JeanPaulLucien** [matrix-org/synapse#15142]
### Description

Could you move this issue? 
https://github.com/vector-im/element-web/issues/24626

### Steps to reproduce

Another URLs without preview:
`https://twitter.com/dshevchenko_biz/status/943829211957092353`

### Homeserver

matrix.org

### Synapse Version

matrix.org

### Installation Method

I don't know

### Database

matrix.org

### Workers

I don't know

### Platform

matrix.org

### Relevant log output

```shell
...
```

### Anything else that would be useful to know?

Matrix.org is homeserver for room and user, and me.

**Comment by clokep** [comment]
I don't understand what the issue is here? Is it a particular URL that isn't previewing? (If so, please provide the URL.) The linked issue mentions something about the `Access-Control-Allow-Origin` header, but I don't see how that relates here? Is it not being returned in one situation?

**Comment by clokep** [comment]
> The linked issue mentions something about the `Access-Control-Allow-Origin` header, but I don't see how that relates here?

The log line was very long, it seems the error is:

```
Access to fetch at 'https://matrix-client.matrix.org/_matrix/media/r0/preview_url?url=https%3A%2F%2Fodysee.com%2F%40Neroke5%3A8&ts=1674980040000'
from origin 'vector://vector' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present
on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors'
to fetch the resource with CORS disabled.
```

Still it isn't clear -- does this always happen with this specific URL? If there are exactly 3 URLs?

**Comment by JeanPaulLucien** [comment]
No preview for this URL always. Room where this URL without preview is public. Other error in Element Desktop 1.11.23 now:
`rageshake.ts:73 Failed to get URL preview: ConnectionError: fetch failed: Failed to fetch`

**Comment by clokep** [comment]
Pulling the logs from that it seems to be:

```
2023-02-22 10:11:33,605 - synapse.http.server - 107 - INFO - GET-79d6f0304927d94f-HEL - <XForwardedForRequest at 0x7fef97109be0 method='GET' uri='/_matrix/media/r0/preview_url?url=https%3A%2F%2Fodysee.com%2F%40Neroke5%3A8&ts=1674980040000' clientproto='HTTP/1.1' site='19103'> SynapseError: 502 - Requested file's content type not allowed for this operation: text/plain; charset=utf-8
2023-02-22 10:11:33,607 - synapse.access.http.19103 - 460 - INFO - GET-79d6f0304927d94f-HEL - XXX.XXX.XXX.XXX - 19103 - {XXX} Processed request: 1.032sec/0.002sec (0.001sec, 0.000sec) (0.000sec/0.000sec/0) 121B 502 "GET /_matrix/media/r0/preview_url?url=https%3A%2F%2Fodysee.com%2F%40Neroke5%3A8&ts=1674980040000 HTTP/1.1" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Element/1.11.23 Chrome/108.0.5359.215 Electron/22.2.0 Safari/537.36" [0 dbevts]
```

The first log line is interesting though -- it seems we're rejecting it due to the content type. This comes from:

https://github.com/matrix-org/synapse/blob/d0fed7a37b8b6ce166cae856fe243757aa7c7294/synapse/http/client.py#L752-L761

We call that via:

https://github.com/matrix-org/synapse/blob/682151a464f688768d5bd8308e16bd4024ad2e57/synapse/rest/media/v1/preview_url_resource.py#L469-L483

This all hinges on the `_is_previewable` function:

https://github.com/matrix-org/synapse/blob/682151a464f688768d5bd8308e16bd4024ad2e57/synapse/rest/media/v1/preview_url_resource.py#L852-L871

The tl;dr is that for some (?) requests to this URL we're getting back a content type of `text/plain`. I tried this locally though and got HTML back, so not sure what's going on there.

----

We should be passing back the proper headers with this error though, I think? But realistically it is still an error response.

**Comment by JeanPaulLucien** [comment]
No preview for URLs that do not exist (404 error). It means when this issue will be resolved we can see a difference between good and others links. This is the  practice, UX effect.

**Comment by clokep** [comment]
> No preview for URLs that do not exist (404 error). It means when this issue will be resolved we can see a difference between good and others links. This is the practice, UX effect.

The [spec for previews](https://spec.matrix.org/v1.5/client-server-api/#get_matrixmediav3preview_url) provides no way to give error information back like that, and I think clients would still just not show a preview.

I'm not sure there's any action for the Synapse team to take here. (We should fix the `Access-Control-Allow-Origin` header, but that's not the root problem.)
